### PR TITLE
AKV: Bring changes from two recent PRs to 7.2-preview into 7.3-preview

### PR DESCRIPTION
**Pull request by daviddesberg**
Bringing in custom role definitions as well as a fix to KeyOperationResult that were recently merged only into 7.2-preview and not 7.3-preview. 

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+    "parameters": {
+        "vaultBaseUrl": "https://myvault.vault.azure.net/",
+        "scope": "",
+        "api-version": "7.2-preview",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json:5]: You should re-record these to make sure all data is right. This is for 7.2-preview, which you're putting in the 7.3-preview folder.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/GetRoleDefinition-example.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+    "parameters": {
+        "vaultBaseUrl": "https://myvault.vault.azure.net/",
+        "scope": "",
+        "api-version": "7.2-preview",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/GetRoleDefinition-example.json:5]: Ditto

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/PutRoleDefinition-example.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            "type": "CustomRole",
+            "description": "Role description",
+            "permissions": [
+                {
+                    "dataActions": [ "Microsoft.KeyVault/managedHsm/keys/sign/action" ],
+                }
+            ]
+        }
+      },
+      "api-version": "7.2-preview"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/PutRoleDefinition-example.json:18]: Ditto

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/rbac.json [hunk 4]
```diff
@@ -365,6 +514,18 @@
...
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/RoleDefinitionProperties",
+          "description": "Role definition properties."
+        }
+      },
+      "required": [
+        "properties"
+      ],
+      "description": "Role definition create parameters."
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/rbac.json:527]:
```suggestion
      "description": "Role definition creation parameters."
```

**Review by heaths — APPROVED** [review]

**Review by herveyw-msft — APPROVED** [review]

**Review by heaths — APPROVED** [review]

**Review by jhendrixMSFT — APPROVED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                "properties": {
+                    "roleName": "My custom role",
+                    "type": "CustomRole",
+                    "description": "Role description",
+                    "assignableScopes": [
+                        "/"
+                    ],
+                    "permissions": [
+                        {
+                            "dataActions": [ "Microsoft.KeyVault/managedHsm/keys/sign/action" ],
```
Comment (jhendrixMSFT) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json:20]: Trailing comma is invalid.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
Causing parsing error.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json (hunk 5 above)
Comment (Lusitanian) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json]: thanks, fixed for all role def examples

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
+{
+    "parameters": {
+        "vaultBaseUrl": "https://myvault.vault.azure.net/",
+        "scope": "",
```
Comment (jhendrixMSFT) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/DeleteRoleDefinition-example.json:4]: Empty value causes validation to fail.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/GetRoleDefinition-example.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+    "parameters": {
+        "vaultBaseUrl": "https://myvault.vault.azure.net/",
+        "scope": "",
```
Comment (jhendrixMSFT) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.3-preview/examples/GetRoleDefinition-example.json:4]: Empty value causes validation to fail.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
Few minor fixes to new examples.

**Review by heaths — APPROVED** [review]

**Comment by jhendrixMSFT** [comment]
There are merge conflicts which need to be resolved.

Also can you please fix the model validation and prettier check CI failures.

**Review by heaths — APPROVED** [review]

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by heaths** [comment]
@Lusitanian the problem is two instances of this:
> In operation "RoleDefinitions_Delete", example for parameter "scope": "/" starts with a forward slash and the path template: "/{scope}/providers/Microsoft.Authorization/roleDefinitions/{roleDefinitionName}" contains a forward slash before the parameter starts. This will cause double forward slashes in the request url. Thus making it incorrect. Please rectify the example.
Url: Microsoft.KeyVault/preview/7.3-preview/rbac.json#L38

I get what's it's saying. Looking at examples, there's one of just "keys" and then one like "/", while "{scope}" should be "/" or "/keys", thereby making the path template "{scope}/providers/...". Does that work? How does ARM handle this? We should probably be copying them.

**Comment by Lusitanian** [comment]
> @Lusitanian the problem is two instances of this:
> 
> > In operation "RoleDefinitions_Delete", example for parameter "scope": "/" starts with a forward slash and the path template: "/{scope}/providers/Microsoft.Authorization/roleDefinitions/{roleDefinitionName}" contains a forward slash before the parameter starts. This will cause double forward slashes in the request url. Thus making it incorrect. Please rectify the example.
> > Url: Microsoft.KeyVault/preview/7.3-preview/rbac.json#L38
> 
> I get what's it's saying. Looking at examples, there's one of just "keys" and then one like "/", while "{scope}" should be "/" or "/keys", thereby making the path template "{scope}/providers/...". Does that work? How does ARM handle this? We should probably be copying them.

In ARM near as I can tell the root scope ("/") is rarely if ever used in URIs so it's difficult to say.  That said, the service is not picky and will accept "//" or "/" for the root scope. I don't think this is a bug.

**Comment by heaths** [comment]
@jhendrixMSFT is it possible we can exempt this error? I know I couldn't with the plethora of `MISSING_REQUIRED_PARAMETER` errors due to a bug in the validation. Or can it just be ignored since it isn't a problem?

David and I discussed it offline and alternatives don't seem like good UX. It either means taking "/" out of the scope, which is inconsistent with ARM when using scopes like "keys", or accepting an empty string meaning "/", which feels like magic. SDKs could normalize this, but we'd prefer the swagger work for any client whether we generated it (and customized it) or not.

**Comment by jhendrixMSFT** [comment]
I don't know if this type of failure can be suppressed, I will check.  However, the error message states "Please rectify the example." which makes me wonder if changing the examples to "/foo" or the like is sufficient?

**Comment by heaths** [comment]
That could work just to satisfy the error. We'll just hit this again when re-recorded, though.

**Comment by heaths** [comment]
Can we force merge this? This validation is wrong in this case - the URL will be fine, as tested by both the service and client teams.

/cc @lmazuel who's been on a few threads/PRs coming in hot from Key Vault and Managed HSM.

**Comment by jhendrixMSFT** [comment]
From the conversation is sounds to me that the `DOUBLE_FORWARD_SLASHES_IN_URL` check should only be applicable to ARM.  I've opened https://github.com/Azure/oav/issues/567 to track getting that fixed (please speak up now if anybody disagrees).

In the meantime I will work on getting this merged.

**Pull request merged**